File: Miscellaneous/credit.py
def get_card_type(card_number: int) -> str:
    # Convert the card number to a string for easy processing
    card_str = str(card_number)
    
    # Step 1: Implement Luhn's algorithm to validate the card number
    sum_digits = 0
    count = len(card_str)
    
    # Traverse the digits from right to left
    for i in range(count - 1, -1, -1):
        digit = int(card_str[i])
        # Even index (counting from 0) for the Luhn algorithm: double the digit and add the sum of digits if the result is >= 10
        if (count - i) % 2 == 0:
            product = digit * 2
            if product >= 10:
                sum_digits += product % 10 + product // 10
            else:
                sum_digits += product
        else:
            sum_digits += digit
    
    # If sum of digits modulo 10 is not zero, the card is invalid
    if sum_digits % 10 != 0:
        return "INVALID"
    
    # Step 2: Determine card type based on the number of digits and starting digits
    # Get the first two digits of the card number
    first_two_digits = int(card_str[:2])
    
    # Determine the card type based on length and starting digits
    if count == 13 and card_str[0] == '4':
        return "VISA"
    elif count == 16:
        if first_two_digits >= 51 and first_two_digits <= 55:
            return "MASTERCARD"
        elif card_str[0] == '4':
            return "VISA"
    elif count == 15 and (first_two_digits == 34 or first_two_digits == 37):
        return "AMEX"
    
    # If none of the conditions match, return INVALID
    return "INVALID"

File: Miscellaneous/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Prompt the user for a long credit card number greater than zero.
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num < 0);

    // Duplicate the copy of the number.
    long temp = num;

    // Sum of digits according to Luhn's algorithm, and count is the number of digits.
    int sum = 0;
    int count = 0;
    int product;

    while (num > 0)
    {
        // If the digit is at an even position from right to left.
        if (count % 2 == 0)
        {
            sum += num % 10;
        }

        // If the digit is at an odd position from right to left.
        else
        {
            product = 2 * (num % 10);

            if (product >= 10)
            {
                // if product is two digit then add first and second digit to sum.
                sum += product % 10; // Gives second digit
                sum += 1;            // first digit is alway one
            }

            else
            {
                sum += product;
            }
        }

        count++;
        num = num / 10; // Discard the last digit.
    }

    // Divide temp by 10 (count - 2) times to obtain the first_two_digit.
    for (int i = 0; i < count - 2; i++)
    {
        temp = temp / 10;
    }

    int first_two_digit = temp;

    // printf("Sum %i\n", sum);
    // printf("Count %i\n", count);
    // printf("First two digits: %i\n", first_two_digit);

    // Check if the card number is valid
    if (sum % 10 == 0)
    {
        // Determine card type based on starting digits
        if ((count == 13 || count == 16) && first_two_digit / 10 == 4)
        {
            printf("VISA\n");
        }
        else if (count == 16 && first_two_digit >= 51 && first_two_digit <= 55)
        {
            printf("MASTERCARD\n");
        }
        else if (count == 15 && (first_two_digit == 34 || first_two_digit == 37))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}